pvsite_datamodel/sqlmodels.py

```diff
@@ -264,3 +268,53 @@ class StatusSQL(Base, CreatedMixin):
     status_uuid = sa.Column(UUID(as_uuid=True), default=uuid.uuid4, primary_key=True)
     status = sa.Column(sa.String(255))
     message = sa.Column(sa.String(255))
+
+
+class InverterSQL(Base, CreatedMixin):
+    """Class representing the inverters table.
+
+    Each InverterSQL row represents an inverter identified in
+    a successful Enode link.
+
+    *Approximate size: *
+    4 clients * ~1000 sites each * ~1 inverter each = ~4000 rows
+    """
+
+    __tablename__ = "inverters"
+
+    inverter_uuid = sa.Column(UUID(as_uuid=True), default=uuid.uuid4, primary_key=True)
+    client_uuid = sa.Column(
```

Comment:
Our thought was that we need to point to both client and site because a user may have linked some inverters from their vendor accounts that they haven't yet attached to a site, but must still maintain a relationship in the database. 

> a client has multiple sites and each site has multiple inverters

This is the precise relationship. Forecasts and Generation data will still be associated with a site in this model. The reason we need a distinction between an inverter and a site is because a site might have multiple, for example, in the case of Enphase's microinverter system. This is not an uncommon system for users of `pv-sites-mobile`. The reason this table doesn't contain other data is because the UUID here comes from Enode, which can be used with Enode to get more data.

Would it be more clear if we marked this table as related to Enode by changing the name to something like enode_inveters?

Reply:
Do I understand correctly that in Enode, there is generation data per inverter, but in `pv-site`, the generation data is aggregated per site?

If so, why can't the service filling in the generation data do:
* Ask Enode for the data of the inverters of a given site
* Aggregate that data (we end up with one value per site per timestamp)
* Save that data in the OCF database
* Then we get forecasts per site

I still don't understand why we need any inverter data in the `pv-site` database, if we end up aggregating everything per site anyway. Sorry if I'm missing something obvious 😅

Reply:
Thanks @AndrewLester this is starting to make a lot of sense!

Here is what I think is a summary of the state of things, along with some ideas.
Please let me know if I still understand some things wrong!

@peterdudfield, you also probably want to have a look!


## Inverter VS site
First I think Enode's `inverter` best aligns with OCF's `site`. It's "something" for which we have a power time series and can make a power forecast. We already assume that the client will have their own id for this `site` and that's what Enode's inverter_uuid would be.
Aggregation could probably happen inside Enode, either before the power generation is sent to us (in this case the `site` would be the sum of many `inverters`), or after the prediction (in this case `site`=`inverter` and the user can aggregate the predictions to get the total forecast of their `sites`/`inverters`).

## Current setup
Our current DB schema is made to accommodate the use-case where a company (the `client`) has many `sites`; each `site` has an associated time series and we are simply trying to forecast that time series.

Here a `site` could be either a panel, an array of panels, an inverter, etc. It can have some metadata like location, orientation, capacity, etc. but we've seen that those don't always make sense depending of what the device producing the time series really is.

We are in the process of adding `users` but so far the users are only thought as "people associated to a given client that manage *all the sites of the client*". In other words `users` are employees of the `client`: they are the ones managing the `sites` and querying the forecasts. It's not clear if we even need a `user` table: we might be able to deal with those inside auth0 altogether.

In this scenario, if the `client` has their own concept of `user`, and want to show a subset of the `client`'s `sites` to specific `users`, the `client` would have to build that extra layer themselves. 

**Definitions in this generic setup**
`client`: a company
`site`: some device associated to a `client` for which we have a time series
`user`: an employee of the `client` that can manage its `sites`

## Enode use-case
My understanding is that the only major difference with the Enode use-case is that `users` only have access to their own `sites`. They are not employees of Enode but users of Enode. This complexifies the permissions a lot: we would need resource-based permissions for the `sites`.

**Definitions in this use-case**
`client`: Enode
`site`: an inverter of a `user` of Enode
`user`: user of Enode that manages their own `sites`

An other option is to define the `client` not as "Enode" but as the user themself. One `client` with a single `user` that can manage their own `sites`.

**Definitions if the Enode user is the client**
`client`: some user of Enode
`site`: inverter for a given Enode user
`user`: same as the client; in this case we only have one `user` per `client`

## Summary

The question to me is: should we support end users of a client managing sites for those end users, or should we just support the employees of that client and let them deal with the end users.

That feels like a way more involved product that what we currently have and would require quite a bit of designing.

An solution could be to treat enode's users as full-blown clients, which would fit our current paradigm where users of a client can manage all the sites of the client.

Reply:
Really appreciate the detailed overview you provided here @simlmx! I think your description of the client vs. user distinction is something we were looking to understand better. I’m commenting this in regard to your above comment and the one you left in the other thread.

@ericcccsliu discussed this a bit, and we had a few thoughts. To recap our proposed implementation:
- For [pv-sites-mobile](https://github.com/openclimatefix/pv-sites-mobile) submitted sites, we are considering one SiteSQL record to be a single panel or cluster of panels in the same general location with the same direction and orientation (similar to [dbdocs definition](https://dbdocs.io/ocf/pvsite?table=sites&schema=public&view=table_structure#:~:text=Each%20site%20row%20specifies%20a%20single%20panel%20or%20cluster%20of%20panels%20found%20on%20a%20residential%20house%20or%20commercial%20building.%20Their%20data%20is%20provided%20by%20a%20client.)).
- If a user does not connect any inverter(s) for a site, then we consider their site’s metadata (including `inverter_capacity_kw`) as important for forecasting, and thus collect the data on a site level. They’ll sum the capacity(ies) of their (unlinked to Enode) inverter(s) for the site, and provide that as the `inverter_capacity_kw`.
- If a user does connect their inverter(s) for this site, and if their site is setup with microinverters, this would mean multiple inverters are connected to this one SiteSQL record. The `inverter_capacity_kw` would no longer be as valuable for forecasting (if I’m correct), but would be useful for calculating clearsky output. The inverters themselves would have their output summed to provide a generation value for the site.

Now, onto the alternatives:
If we tried to avoid this, we would have multiple SiteSQL records for identical metadata and similar time series. In the case where a user has several microinverters for a single site, we could create an individual SiteSQL record for each inverter, given it has its own timeseries.
- However, given that the metadata for each of the SiteSQL records would be identical, the timeseries will also be similar (given that similar metadata is a good predictor of similar forecasts, which we think is true in the individual homeowner case, but feel free to comment on this).
- The extra timeseries would require more processing time on the forecasting side and would require more storage for both forecasting and generation.
- An equivalent, and equally difficult setup would be to have an individual SiteSQL record for each panel. This does not appear to be what is done currently for sites, and therefore it is probably not what we want to do when multiple inverters can be linked.
- Aggregating inverter data and performing a forecast or storing the aggregate as generation data is just doing what is already done for inputting `site_capacity_kw` in a site with more than one panel, where the panel nameplate capacities are summed.

Let me know if I'm incorrect in any of the points I mentioned here, but I think given this last set of information we can come to a consensus.